### PR TITLE
Allow perpetual dft mining to correctly accept the next bitwork

### PR DESCRIPTION
**Pull request by atomicals**
**Problem**

Users who mine the next bitwork difficulty instead of the current will not be awarded the dft mint. This problem is magnified due to the atomicals-js CLI as of the previous version choosing to mine the next bitwork difficulty for an infinitely mineable token.

The user impact is that the users who attempted to mine the next difficulty after a certain threshold will not get their token mint because the indexer does not recognize bitwork difficulties higher than the currently acceptable one.

For example, if the current bitwork is `7777.15` and the next bitwork is `77777` then anyone who mined `77777` will not be accepted because according to the current rules (before this fix) only `7777f` is valid (due to the presence of `15`). 

The solution below focuses on also allowing the `77777` to be mined along with the `7777f` since they are both technically the same difficulty. 

**Solution**

The solution presented here is to allow the indexer flexibility to also accept the *next* difficulty gracefully and award the user their dft mint.

The solution is activated (param `ATOMICALS_ACTIVATION_HEIGHT_BITWORKEXT`) on livenet as of height `828500` and testnet as of height  `2576412`.

When the new configuration becomes active, a Bitwork will be accepted from the next increment roll over. 

Unit tests:

*Note*: Care must be taken for roll overs with increments that are not necessarily evenly divided by 16.
*Run tests with*: `pytest  ./tests/lib/test_atomicals_utils.py`

https://github.com/atomicals/atomicals-electrumx/blob/accept-next-bitwork/tests/lib/test_atomicals_utils.py#L463

### CODE REVIEW

File: tests/lib/test_atomicals_utils.py [hunk 1]
```diff
@@ -459,3 +460,116 @@ def test_calculate_expected_bitwork_base():
     assert(calculate_expected_bitwork('abcdefe', 33000, 1000, 1, 127) == 'abcdefe000')
     assert(calculate_expected_bitwork('abcdefe', 33000, 1000, 3, 127) == 'abcdefe0000000.2')
      
+def test_calculate_expected_bitwork_rollover():
+    
+    assert(calculate_expected_bitwork('888888888888', 49995, 3333, 1, 64) == '8888.15')
```
Comment (AlexV525) [inline tests/lib/test_atomicals_utils.py:465]: Consider adding cases to guard increments other than `1`.

**Review by AlexV525 — COMMENTED** [review]
The idea is good overall. Thanks for the patch which significantly helps users to save their mints!

To prove the `next` difficulty can be equal to the `current`, we might make tests to iterate the combination of `inc` and `start` to find out if we can avoid anti-patterns as much as possible.

File: electrumx/lib/util_atomicals.py [hunk 2]
```diff
@@ -1664,6 +1663,21 @@ def get_subname_request_candidate_status(current_height, atomical_info, status,
         'pending_candidate_atomical_id': candidate_id_compact
     }
 
+# Whether txid is valid for the current and next bitwork
+def is_txid_valid_for_bitwork(txid, bitwork_vec, actual_mints, max_mints, target_increment, starting_target, allow_next):
```
Comment (AlexV525) [inline electrumx/lib/util_atomicals.py:1667]: For readability: rename to `is_txid_valid_for_perpetual_bitwork`

File: electrumx/server/block_processor.py [hunk 3]
```diff
@@ -2572,19 +2576,30 @@ def create_or_delete_decentralized_mint_output(self, atomicals_operations_found_
...
                  # If there was a reveal bitwork required, then assess the stage of the minimum we expect to allow the mint
                 if mint_bitworkr_inc:
-                    mint_bitworkr_start = mint_info_for_ticker.get('$mint_bitworkr_start')  
-                    expected_minimum_bitworkr = calculate_expected_bitwork(mint_bitwork_vec, decentralized_mints, max_mints, mint_bitworkr_inc, mint_bitworkr_start)
-                    if not is_mint_pow_valid(atomicals_operations_found_at_inputs['reveal_location_txid'], expected_minimum_bitworkr):
-                        self.logger.warning(f'create_or_delete_decentralized_mint_output: mint_bitworkr_inc not is_mint_pow_valid {hash_to_hex_str(tx_hash)}, expected_minimum_bitworkr={expected_minimum_bitworkr}, atomicals_operations_found_at_inputs={atomicals_operations_found_at_inputs}...')
-                        return None   
-
+                    mint_bitworkr_start = mint_info_for_ticker.get('$mint_bitworkr_start')
+                    if height >= self.coin.ATOMICALS_ACTIVATION_HEIGHT_BITWORKEXT:
```
Comment (shadowv0vshadow) [inline electrumx/server/block_processor.py:2594]: Get error `AttributeError: type object 'BitcoinTestnet' has no attribute 'ATOMICALS_ACTIVATION_HEIGHT_BITWORKEXT'`

**Comment by jiedo** [comment]
Don't just fix the problem for the cli mining mode, we need to make the txid that already meets the longer bv prefix valid under any shorter bitwork requirements with ext.

File: electrumx/server/block_processor.py (hunk 3 above)
Comment (NeutronProtocol) [inline electrumx/server/block_processor.py:2594]: this is essentially
 ```suggestion
                    if self.is_dft_bitwork_rollover_activated(height):
```

File: tests/lib/test_atomicals_utils.py (hunk 1 above)
Comment (NeutronProtocol) [reply tests/lib/test_atomicals_utils.py:465]: please also consider adding cases for more general cases like: 8888.2, 8888.3, 8888.8, 8888.9

**Comment by shadowv0vshadow** [comment]
Fully test, it work, here is my test log
```
txid: 17fc218b106cad775c2a3a8a47c9ef650deaf4f29bd89007b61625c16329bb17
max_mints_global: None
mint_bitwork_vec: 777777777777
mint_bitworkc_inc: 1
mint_bitworkr_inc: 0
height 2576474
mint_bitworkc_start: 64
bitwork_str: 7777.15
success: True
txid: cccb1574a69913de166b68bccd05f079565a1a9c43981d91c702c202538b21b3
max_mints_global: None
mint_bitwork_vec: 777777777777
mint_bitworkc_inc: 1
mint_bitworkr_inc: 0
height 2576474
mint_bitworkc_start: 64
bitwork_str: None
success: False
txid: d783b6c804ea0cff00ced96cc199f01d7ccba5fed282d7560d42e7c46881a37c
max_mints_global: None
mint_bitwork_vec: 777777777777
mint_bitworkc_inc: 1
mint_bitworkr_inc: 0
height 2576477
mint_bitworkc_start: 64
bitwork_str: 77777
success: True
perpetual
txid: dc5ece2dd7b248d289aa4376019165b2a21880d10ed6cbb53ae97cdbba0cb7de
max_mints_global: None
mint_bitwork_vec: 777777777777
mint_bitworkc_inc: 1
mint_bitworkr_inc: 0
height 2576477
mint_bitworkc_start: 64
bitwork_str: 77777
success: True
perpetual
txid: 815ac388f2cb856c01c6aad4dd56656b948bc71872c2f103bb565bad75c2e2f3
max_mints_global: None
mint_bitwork_vec: 777777777777
mint_bitworkc_inc: 1
mint_bitworkr_inc: 0
height 2576478
mint_bitworkc_start: 64
bitwork_str: 77777
success: True
```
The only failed tx's bitwork is `7777b`, Indexer accpeted the 77777 when mint_bitworkc_current equal 7777.15.

File: electrumx/lib/util_atomicals.py (hunk 2 above)
Comment (atomicals) [reply electrumx/lib/util_atomicals.py]: Done

**Comment by atomicals** [comment]
Please review again as there have been some changes and additional tests to accept the next full complete bitwork string.

ie:   88888 should be valid for  8888.10

File: electrumx/lib/util_atomicals.py [hunk 4]
```diff
@@ -1664,6 +1663,28 @@ def get_subname_request_candidate_status(current_height, atomical_info, status,
         'pending_candidate_atomical_id': candidate_id_compact
     }
 
+def get_next_bitwork_full_str(bitwork_vec, current_prefix_len):
+    base_bitwork_padded = bitwork_vec.ljust(32, '0') 
+    b = base_bitwork_padded[:current_prefix_len]
+    p = base_bitwork_padded[:current_prefix_len + 1]
+    print(f'next_full_bitwork_prefix b={b} p={p} ')
```
Comment (shadowv0vshadow) [inline electrumx/lib/util_atomicals.py:1670]: No need print, logger.debug?

File: electrumx/server/block_processor.py [hunk 5]
```diff
@@ -2572,19 +2576,30 @@ def create_or_delete_decentralized_mint_output(self, atomicals_operations_found_
...
+                        if not is_txid_valid_for_perpetual_bitwork(atomicals_operations_found_at_inputs['commit_txid'], mint_bitwork_vec, decentralized_mints, max_mints, mint_bitworkr_inc, mint_bitworkr_start, True):
+                            self.logger.warning(f'create_or_delete_decentralized_mint_output: mint_bitworkc_inc not is_mint_pow_valid {hash_to_hex_str(tx_hash)}, atomicals_operations_found_at_inputs={atomicals_operations_found_at_inputs}...')
```
Comment (AlexV525) [inline electrumx/server/block_processor.py:2596]:
```suggestion
                            self.logger.warning(f'create_or_delete_decentralized_mint_output: mint_bitworkr_inc not is_mint_pow_valid {hash_to_hex_str(tx_hash)}, atomicals_operations_found_at_inputs={atomicals_operations_found_at_inputs}...')
```

File: electrumx/server/block_processor.py [hunk 6]
```diff
@@ -2572,19 +2576,30 @@ def create_or_delete_decentralized_mint_output(self, atomicals_operations_found_
...
+                            return None
+                    else: 
+                        if not is_txid_valid_for_perpetual_bitwork(atomicals_operations_found_at_inputs['reveal_location_txid'], mint_bitwork_vec, decentralized_mints, max_mints, mint_bitworkr_inc, mint_bitworkr_start, False):
+                            self.logger.warning(f'create_or_delete_decentralized_mint_output: mint_bitworkc_inc not is_mint_pow_valid {hash_to_hex_str(tx_hash)}, atomicals_operations_found_at_inputs={atomicals_operations_found_at_inputs}...')
```
Comment (AlexV525) [inline electrumx/server/block_processor.py:2600]:
```suggestion
                            self.logger.warning(f'create_or_delete_decentralized_mint_output: mint_bitworkr_inc not is_mint_pow_valid {hash_to_hex_str(tx_hash)}, atomicals_operations_found_at_inputs={atomicals_operations_found_at_inputs}...')
```

File: electrumx/server/block_processor.py [hunk 7]
```diff
@@ -2572,19 +2576,30 @@ def create_or_delete_decentralized_mint_output(self, atomicals_operations_found_
...
                 if mint_bitworkr_inc:
-                    mint_bitworkr_start = mint_info_for_ticker.get('$mint_bitworkr_start')  
-                    expected_minimum_bitworkr = calculate_expected_bitwork(mint_bitwork_vec, decentralized_mints, max_mints, mint_bitworkr_inc, mint_bitworkr_start)
-                    if not is_mint_pow_valid(atomicals_operations_found_at_inputs['reveal_location_txid'], expected_minimum_bitworkr):
-                        self.logger.warning(f'create_or_delete_decentralized_mint_output: mint_bitworkr_inc not is_mint_pow_valid {hash_to_hex_str(tx_hash)}, expected_minimum_bitworkr={expected_minimum_bitworkr}, atomicals_operations_found_at_inputs={atomicals_operations_found_at_inputs}...')
-                        return None   
-
+                    mint_bitworkr_start = mint_info_for_ticker.get('$mint_bitworkr_start')
+                    if height >= self.coin.ATOMICALS_ACTIVATION_HEIGHT_BITWORKEXT:
+                        if not is_txid_valid_for_perpetual_bitwork(atomicals_operations_found_at_inputs['commit_txid'], mint_bitwork_vec, decentralized_mints, max_mints, mint_bitworkr_inc, mint_bitworkr_start, True):
```
Comment (AlexV525) [inline electrumx/server/block_processor.py:2595]: Should be `reveal_location_txid`?
```suggestion
                        if not is_txid_valid_for_perpetual_bitwork(atomicals_operations_found_at_inputs['reveal_location_txid'], mint_bitwork_vec, decentralized_mints, max_mints, mint_bitworkr_inc, mint_bitworkr_start, True):
```

Comment (shadowv0vshadow) [reply electrumx/server/block_processor.py]: ATOMICALS_ACTIVATION_HEIGHT_BITWORKEXT not define, need fix

File: electrumx/server/block_processor.py (hunk 3 above)
Comment (atomicals) [reply electrumx/server/block_processor.py]: Done

Comment (atomicals) [reply electrumx/server/block_processor.py]: Done

File: electrumx/lib/util_atomicals.py (hunk 4 above)
Comment (atomicals) [reply electrumx/lib/util_atomicals.py]: Done and removed

File: electrumx/server/block_processor.py (hunk 5 above)
Comment (atomicals) [reply electrumx/server/block_processor.py]: Done

File: electrumx/server/block_processor.py (hunk 6 above)
Comment (atomicals) [reply electrumx/server/block_processor.py]: Done

File: electrumx/server/block_processor.py (hunk 7 above)
Comment (atomicals) [reply electrumx/server/block_processor.py]: Done

File: tests/lib/test_atomicals_utils.py (hunk 1 above)
Comment (atomicals) [reply tests/lib/test_atomicals_utils.py:466]: Done, added a variety and using an array-loop

**Comment by AlexV525** [comment]
FYI to everyone: the activation height is `828628` instead of `828500`.

https://github.com/atomicals/atomicals-electrumx/blob/2372c5680dfe924e91cf30c25de46bdb7b51fbc7/electrumx/lib/coins.py#L669

**Pull request merged**